File: Go/PollingSystem2-go/repositories/vote_repository.go
package repositories

import (
	"fmt"
	"sync"

	"lld-practice/pollingsystem2-go/models"
)

type VoteRepository struct {
	mu    sync.RWMutex
	votes []*models.Vote
}

func NewVoteRepository() *VoteRepository {
	return &VoteRepository{votes: []*models.Vote{}}
}

func (r *VoteRepository) Add(vote *models.Vote) error {
	if vote == nil {
		return fmt.Errorf("invalid vote")
	}
	r.mu.Lock()
	defer r.mu.Unlock()
	for _, v := range r.votes {
		if v.PollID == vote.PollID && v.UserID == vote.UserID {
			return fmt.Errorf("user has already voted on this poll!")
		}
	}
	r.votes = append(r.votes, vote)
	return nil
}

func (r *VoteRepository) GetStatistics(poll *models.Poll) map[string]any {
	r.mu.RLock()
	defer r.mu.RUnlock()

	var filtered []*models.Vote
	for _, v := range r.votes {
		if v.PollID == poll.ID {
			filtered = append(filtered, v)
		}
	}

	ratings := make(map[string]int)
	for _, v := range filtered {
		ratings[v.Option]++
	}

	total := len(filtered)
	stats := make(map[string]float64)
	counts := make(map[string]int)
	for _, option := range poll.Options {
		count := ratings[option]
		counts[option] = count
		if total == 0 {
			stats[option] = 0
		} else {
			stats[option] = float64(count) / float64(total)
		}
	}

	return map[string]any{
		"question":   poll.Question,
		"totalVotes": total,
		"counts":     counts,
		"statistics": stats,
	}
}

File: JavaScript/PollingSystem2/PollingService/repositories/VoteRepository.js
const { Vote } = require('../models/Vote');

/** In-memory stand-in for a vote repository. */
class VoteRepository {
  static votes = [];

  static add(vote) {
    if (!(vote instanceof Vote)) {
      throw new Error('Invalid vote');
    }

    const alreadyVoted = this.votes.some(
      (v) => v.pollId === vote.pollId && v.userId === vote.userId,
    );
    if (alreadyVoted) {
      throw new Error('User has already voted on this poll!');
    }

    this.votes.push(vote);
  }

  static getByPollId(pollId) {
    return this.votes.filter((v) => v.pollId === pollId);
  }

  static getStatistics(poll) {
    const votes = this.getByPollId(poll.id);
    const total = votes.length;

    const ratings = {};
    for (const vote of votes) {
      ratings[vote.option] = (ratings[vote.option] || 0) + 1;
    }

    const statistics = {};
    for (const option of poll.options) {
      const count = ratings[option] || 0;
      statistics[option] = total === 0 ? 0 : count / total;
    }

    return {
      question: poll.question,
      totalVotes: total,
      counts: Object.fromEntries(
        poll.options.map((option) => [option, ratings[option] || 0]),
      ),
      statistics,
    };
  }
}

module.exports = { VoteRepository };
